File: javascript/packages/linter/src/cli/argument-parser.ts
import dedent from "dedent"

import { availableParallelism } from "node:os"
import { parseArgs } from "util"
import { Herb } from "@herb-tools/node-wasm"

import { THEME_NAMES, DEFAULT_THEME } from "@herb-tools/highlighter"
import { DIAGNOSTIC_SEVERITIES, isDiagnosticSeverity } from "@herb-tools/core"

import type { ThemeInput } from "@herb-tools/highlighter"
import type { DiagnosticSeverity } from "@herb-tools/core"

import { name, version, dependencies } from "../../package.json"

export type FormatOption = "simple" | "detailed" | "json"

export interface ParsedArguments {
  patterns: string[]
  configFile?: string
  formatOption: FormatOption
  showTiming: boolean
  theme: ThemeInput
  wrapLines: boolean
  truncateLines: boolean
  useGitHubActions: boolean
  fix: boolean
  fixUnsafe: boolean
  ignoreDisableComments: boolean
  force: boolean
  init: boolean
  upgrade: boolean
  disableFailing: boolean
  loadCustomRules: boolean
  failLevel?: DiagnosticSeverity
  logLevel?: DiagnosticSeverity
  jobs: number
  only?: string[]
  allRules: boolean
}

export class ArgumentParser {
  private readonly usage = dedent`
    Usage: herb-lint [files|directories|glob-patterns...] [options]

    Arguments:
      files            Files, directories, or glob patterns to lint (defaults to configured extensions in .herb.yml)
                       Multiple arguments are supported (e.g., herb-lint file1.erb file2.erb dir/ "**/*.erb")

    Options:
      -h, --help                    show help
      -v, --version                 show version
      --init                        create a .herb.yml configuration file in the current directory
      --upgrade                     update .herb.yml version and disable all newly introduced rules
      --disable-failing             lint the codebase and disable all rules that have offenses in .herb.yml
      -c, --config-file <path>      explicitly specify path to .herb.yml config file
      --force                       force linting even if disabled in .herb.yml
      --only <rules>                only run the given rules, ignoring the rule configuration in .herb.yml
                                    accepts a comma-separated list and can be passed multiple times
                                    (e.g., herb-lint --only html-img-require-alt,html-tag-name-lowercase)
      --all-rules                   run every rule, ignoring the rule configuration in .herb.yml
                                    including rules that are disabled, not enabled by default,
                                    or introduced after the version in .herb.yml
      --fix                         automatically fix auto-correctable offenses
      --fix-unsafely                also apply unsafe auto-fixes (implies --fix)
      --ignore-disable-comments     report offenses even when suppressed with <%# herb:disable %> comments
      --fail-level <severity>       exit with error code when diagnostics of this severity or higher are present (error|warning|info|hint) [default: error]
      --log-level <severity>        only report diagnostics of this severity or higher (error|warning|info|hint) [default: hint]
                                    lower-severity offenses are still counted in the summary, but aren't
                                    printed or annotated in CI
      --format                      output format (simple|detailed|json) [default: detailed]
      --simple                      use simple output format (shortcut for --format simple)
      --json                        use JSON output format (shortcut for --format json)
      --github                      enable GitHub Actions annotations (combines with --format)
      --no-github                   disable GitHub Actions annotations (even in GitHub Actions environment)
      --no-custom-rules             disable loading custom rules from project (custom rules are loaded by default from .herb/rules/**/*.{mjs,js})
      -j, --jobs <n>                number of parallel workers for linting files [default: auto]
                                    use "auto" to detect based on available CPU cores
      --theme                       syntax highlighting theme (${THEME_NAMES.join("|")}) or path to custom theme file [default: ${DEFAULT_THEME}]
      --no-color                    disable colored output
      --no-timing                   hide timing information
      --no-wrap-lines               disable line wrapping
      --truncate-lines              enable line truncation (mutually exclusive with line wrapping)
  `

  parse(argv: string[]): ParsedArguments {
    const { values, positionals } = parseArgs({
      args: argv.slice(2),
      options: {
        help: { type: "boolean", short: "h" },
        version: { type: "boolean", short: "v" },
        init: { type: "boolean" },
        upgrade: { type: "boolean" },
        "disable-failing": { type: "boolean" },
        "config-file": { type: "string", short: "c" },
        force: { type: "boolean" },
        only: { type: "string", multiple: true },
        "all-rules": { type: "boolean" },
        fix: { type: "boolean" },
        "fix-unsafely": { type: "boolean" },
        "ignore-disable-comments": { type: "boolean" },
        "fail-level": { type: "string" },
        "log-level": { type: "string" },
        format: { type: "string" },
        simple: { type: "boolean" },
        json: { type: "boolean" },
        github: { type: "boolean" },
        "no-github": { type: "boolean" },
        theme: { type: "string" },
        "no-color": { type: "boolean" },
        "no-timing": { type: "boolean" },
        "no-wrap-lines": { type: "boolean" },
        "truncate-lines": { type: "boolean" },
        "no-custom-rules": { type: "boolean" },
        jobs: { type: "string", short: "j" }
      },
      allowPositionals: true
    })

    if (values.help) {
      console.log(this.usage)
      process.exit(0)
    }

    if (values.version) {
      console.log("Versions:")
      console.log(`  ${name}@${version}`)
      console.log(`  @herb-tools/printer@${dependencies["@herb-tools/printer"]}`)
      console.log(`  ${Herb.version}`.split(", ").join("\n  "))
      process.exit(0)
    }

    const isGitHubActions = process.env.GITHUB_ACTIONS === "true"

    let formatOption: FormatOption = "detailed"
    if (values.format && (values.format === "detailed" || values.format === "simple" || values.format === "json")) {
      formatOption = values.format
    }

    if (values.simple) {
      formatOption = "simple"
    }

    if (values.json) {
      formatOption = "json"
    }

    const useGitHubActions = (values.github || isGitHubActions) && !values["no-github"]

    if (useGitHubActions && formatOption === "json") {
      console.error("Error: --github cannot be used with --json format. JSON format is already structured for programmatic consumption.")
      process.exit(1)
    }

    if (values["no-color"]) {
      process.env.NO_COLOR = "1"
    }

    const showTiming = !values["no-timing"]

    let wrapLines = !values["no-wrap-lines"]
    let truncateLines = false

    if (values["truncate-lines"]) {
      truncateLines = true
      wrapLines = false
    }

    if (!values["no-wrap-lines"] && values["truncate-lines"]) {
      console.error("Error: Line wrapping and --truncate-lines cannot be used together. Use --no-wrap-lines with --truncate-lines.")
      process.exit(1)
    }

    const theme = values.theme || DEFAULT_THEME
    const patterns = this.getFilePatterns(positionals)
    const fixUnsafe = values["fix-unsafely"] || false
    const fix = values.fix || fixUnsafe  // --fix-unsafely implies --fix
    const force = !!values.force
    const ignoreDisableComments = values["ignore-disable-comments"] || false
    const configFile = values["config-file"]
    const init = values.init || false
    const upgrade = values.upgrade || false
    const disableFailing = values["disable-failing"] || false
    const loadCustomRules = !values["no-custom-rules"]

    const allRules = values["all-rules"] || false

    let only: string[] | undefined

    if (values.only) {
      only = [...new Set(values.only.flatMap(value => value.split(",")).map(ruleName => ruleName.trim()).filter(Boolean))]

      if (only.length === 0) {
        console.error(`Error: --only requires at least one rule name.`)
        process.exit(1)
      }

      if (allRules) {
        console.error(`Error: --only and --all-rules can't be combined.`)
        process.exit(1)
      }
    }

    const failLevel = this.parseSeverity(values["fail-level"], "--fail-level")
    const logLevel = this.parseSeverity(values["log-level"], "--log-level")

    let jobs = availableParallelism()

    if (values.jobs && values.jobs !== "auto") {
      const parsed = parseInt(values.jobs, 10)

      if (isNaN(parsed) || parsed < 1) {
        console.error(`Error: Invalid --jobs value "${values.jobs}". Must be a positive integer or "auto".`)
        process.exit(1)
      }

      jobs = parsed
    }

    return { patterns, configFile, formatOption, showTiming, theme, wrapLines, truncateLines, useGitHubActions, fix, fixUnsafe, ignoreDisableComments, force, init, upgrade, disableFailing, loadCustomRules, failLevel, logLevel, jobs, only, allRules }
  }

  private parseSeverity(value: string | undefined, flag: string): DiagnosticSeverity | undefined {
    if (!value) return undefined

    if (!isDiagnosticSeverity(value)) {
      console.error(`Error: Invalid ${flag} value "${value}". Must be one of: ${DIAGNOSTIC_SEVERITIES.join(", ")}`)
      process.exit(1)
    }

    return value
  }

  private getFilePatterns(positionals: string[]): string[] {
    return positionals
  }
}

File: rust/herb-linter/src/cli/argument_parser.rs
use crate::*;

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum OutputFormat {
  Simple,
  Detailed,
  Json,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) enum FailLevel {
  Error,
  Warning,
  Info,
  Hint,
}

impl FailLevel {
  pub(crate) fn severity(&self) -> Severity {
    match self {
      FailLevel::Error => Severity::Error,
      FailLevel::Warning => Severity::Warning,
      FailLevel::Info => Severity::Info,
      FailLevel::Hint => Severity::Hint,
    }
  }
}

#[derive(Debug)]
#[allow(dead_code)]
pub(crate) struct CliArguments {
  pub(crate) patterns: Vec<String>,
  pub(crate) format: OutputFormat,
  pub(crate) fail_level: FailLevel,
  pub(crate) fail_level_explicit: bool,
  pub(crate) config_file: Option<String>,
  pub(crate) fix: bool,
  pub(crate) fix_unsafe: bool,
  pub(crate) ignore_disable_comments: bool,
  pub(crate) force: bool,
  pub(crate) init: bool,
  pub(crate) github_actions: bool,
  pub(crate) no_color: bool,
  pub(crate) show_timing: bool,
  pub(crate) show_rules: bool,
  pub(crate) upgrade: bool,
  pub(crate) disable_failing: bool,
  pub(crate) show_fix_diff: bool,
  pub(crate) only_rules: Vec<String>,
  pub(crate) all_rules: bool,
  pub(crate) log_level: Option<FailLevel>,
}

pub(crate) fn split_inline_values(arguments: impl Iterator<Item = String>) -> Vec<String> {
  let mut values = Vec::new();

  for argument in arguments {
    match argument.split_once('=') {
      Some((name, value)) if name.starts_with("--") && !name.is_empty() => {
        values.push(name.to_string());
        values.push(value.to_string());
      }
      _ => values.push(argument),
    }
  }

  values
}

pub(crate) fn parse_arguments() -> CliArguments {
  let argument_values: Vec<String> = split_inline_values(std::env::args());

  let mut patterns = Vec::new();
  let mut format = OutputFormat::Detailed;
  let mut fail_level = FailLevel::Error;
  let mut fail_level_explicit = false;
  let mut config_file = None;
  let mut fix = false;
  let mut fix_unsafe = false;
  let mut ignore_disable_comments = false;
  let mut force = false;
  let mut init = false;
  let mut github_actions_flag: Option<bool> = None;
  let mut no_color = false;
  let mut show_timing = true;
  let mut show_fix_diff = false;
  let mut only_rules: Vec<String> = Vec::new();
  let mut all_rules = false;
  let mut log_level: Option<FailLevel> = None;
  let mut show_rules = false;
  let mut upgrade = false;
  let mut disable_failing = false;

  let mut index = 1;
  while index < argument_values.len() {
    let argument = &argument_values[index];

    match argument.as_str() {
      "-h" | "--help" => {
        print_usage();
        std::process::exit(0);
      }
      "-v" | "--version" => {
        println!("herb-lint {}", env!("CARGO_PKG_VERSION"));
        std::process::exit(0);
      }
      "--init" => init = true,
      "--upgrade" => upgrade = true,
      "--disable-failing" => disable_failing = true,
      "-c" | "--config-file" => {
        index += 1;
        if index >= argument_values.len() {
          eprintln!("Error: --config-file requires a path argument");
          std::process::exit(1);
        }
        config_file = Some(argument_values[index].clone());
      }
      "--force" => force = true,
      "--fix" => fix = true,
      "--fix-unsafely" => {
        fix_unsafe = true;
        fix = true;
      }
      "--ignore-disable-comments" => ignore_disable_comments = true,
      "--fail-level" => {
        index += 1;
        if index >= argument_values.len() {
          eprintln!("Error: --fail-level requires a severity argument (error|warning|info|hint)");
          std::process::exit(1);
        }
        fail_level_explicit = true;
        fail_level = match argument_values[index].as_str() {
          "error" => FailLevel::Error,
          "warning" => FailLevel::Warning,
          "info" => FailLevel::Info,
          "hint" => FailLevel::Hint,
          other => {
            eprintln!("Error: Invalid --fail-level value \"{}\". Must be one of: error, warning, info, hint", other);
            std::process::exit(1);
          }
        };
      }
      "--format" => {
        index += 1;
        if index >= argument_values.len() {
          eprintln!("Error: --format requires a format argument (simple|detailed|json)");
          std::process::exit(1);
        }
        format = match argument_values[index].as_str() {
          "simple" => OutputFormat::Simple,
          "detailed" => OutputFormat::Detailed,
          "json" => OutputFormat::Json,
          other => {
            eprintln!("Error: Invalid --format value \"{}\". Must be one of: simple, detailed, json", other);
            std::process::exit(1);
          }
        };
      }
      "--simple" => format = OutputFormat::Simple,
      "--json" => format = OutputFormat::Json,
      "--github" => github_actions_flag = Some(true),
      "--no-github" => github_actions_flag = Some(false),
      "--no-color" => no_color = true,
      "--no-timing" => show_timing = false,
      "--all-rules" => all_rules = true,
      "--show-fix-diff" => show_fix_diff = true,
      "--only" => {
        index += 1;

        match argument_values.get(index) {
          Some(value) => only_rules.extend(value.split(',').map(|name| name.trim().to_string()).filter(|name| !name.is_empty())),
          None => {
            eprintln!("Error: --only requires a comma separated list of rule names");
            std::process::exit(1);
          }
        }
      }
      "--log-level" => {
        index += 1;

        match argument_values.get(index).map(|value| value.as_str()) {
          Some("error") => log_level = Some(FailLevel::Error),
          Some("warning") => log_level = Some(FailLevel::Warning),
          Some("info") => log_level = Some(FailLevel::Info),
          Some("hint") => log_level = Some(FailLevel::Hint),
          Some(other) => {
            eprintln!("Error: Invalid --log-level value \"{}\". Must be one of: error, warning, info, hint", other);
            std::process::exit(1);
          }
          None => {
            eprintln!("Error: --log-level requires a severity argument (error|warning|info|hint)");
            std::process::exit(1);
          }
        }
      }
      "--no-wrap-lines" | "--truncate-lines" | "--theme" => {
        if argument == "--theme" {
          index += 1;
        }
      }
      "--no-custom-rules" => {
        // Accept but ignore
      }
      "rules" if index == 1 => show_rules = true,
      _ if argument.starts_with('-') => {
        eprintln!("Error: Unknown option: {}", argument);
        eprintln!("Run with --help for usage information.");
        std::process::exit(1);
      }
      _ => patterns.push(argument.clone()),
    }
    index += 1;
  }

  let is_github_environment = std::env::var("GITHUB_ACTIONS").map_or(false, |value| value == "true");

  let github_actions = match github_actions_flag {
    Some(explicit) => explicit,
    None => is_github_environment,
  };

  if github_actions && format == OutputFormat::Json {
    eprintln!("Error: --github cannot be used with --json format. JSON format is already structured for programmatic consumption.");
    std::process::exit(1);
  }

  if no_color {
    std::env::set_var("NO_COLOR", "1");
  }

  CliArguments {
    patterns,
    format,
    fail_level,
    fail_level_explicit,
    config_file,
    fix,
    fix_unsafe,
    ignore_disable_comments,
    force,
    init,
    github_actions,
    no_color,
    show_timing,
    show_rules,
    upgrade,
    disable_failing,
    show_fix_diff,
    only_rules,
    all_rules,
    log_level,
  }
}

pub(crate) fn print_usage() {
  println!("herb-lint {} - Linter for HTML+ERB templates", env!("CARGO_PKG_VERSION"));
  println!();
  println!("Usage: herb-lint [files|directories...] [options]");
  println!();
  println!("Arguments:");
  println!("  files            Files, directories, or patterns to lint");
  println!("                   Multiple arguments are supported (e.g., herb-lint file1.erb dir/)");
  println!();
  println!("Commands:");
  println!("  rules            List all available lint rules");
  println!();
  println!("Options:");
  println!("  -h, --help                    show help");
  println!("  -v, --version                 show version");
  println!("  -c, --config-file <path>      explicitly specify path to config file");
  println!("  --force                       force linting even if disabled in config");
  println!("  --fix                         automatically fix auto-correctable offenses");
  println!("  --fix-unsafely                also apply unsafe auto-fixes (implies --fix)");
  println!("  --ignore-disable-comments     report offenses even when suppressed with herb:disable comments");
  println!("  --fail-level <severity>       exit with error code for this severity or higher (error|warning|info|hint) [default: error]");
  println!("  --format <format>             output format (simple|detailed|json) [default: detailed]");
  println!("  --simple                      use simple output format (shortcut for --format simple)");
  println!("  --json                        use JSON output format (shortcut for --format json)");
  println!("  --github                      enable GitHub Actions annotations");
  println!("  --no-github                   disable GitHub Actions annotations (even in GitHub Actions environment)");
  println!("  --no-color                    disable colored output");
  println!("  --no-timing                   hide timing information");
  println!("  --only <rules>                run only the given comma separated rules");
  println!("  --upgrade                     update the .herb.yml version, enabling new rules that have no offenses");
  println!("  --disable-failing             disable every rule that currently reports an offense in .herb.yml");
  println!("  --all-rules                   run every rule, ignoring config and defaults");
  println!("  --show-fix-diff               show what `--fix` would change, without writing");
  println!("  --log-level <severity>        hide offenses below this severity (error|warning|info|hint)");
}
